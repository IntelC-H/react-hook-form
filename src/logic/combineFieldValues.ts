import { FieldValue } from '../types';
import set from '../utils/set';

export default function combineFieldValues(data): FieldValue {
  const output = Object.entries(data).reduce(
    (previous, [key, value]): FieldValue => {
      const arrayIndex = key.match(/\[\d+\]/gi);
      const dotPosition = key.indexOf('.');

      if (arrayIndex || dotPosition) {
        set(previous, key, value);
        return previous;
      }

      previous[key] = value;
      return previous;
    },
    {},
  );

  return Object.entries(output).reduce((previous, [key, value]): FieldValue => {
    if (Array.isArray(value)) {
      previous[key] = value.filter(Boolean);
      return previous;
    }

    previous[key] = value;
    return previous;
  }, {});
}
