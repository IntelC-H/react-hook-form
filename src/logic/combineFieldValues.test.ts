import combineFieldValues from './combineFieldValues';

describe('combineFieldValues', () => {
  it('should combine all the array fields', () => {
    expect(
      combineFieldValues({
        'email[1]': 'asdasd@dsad.com',
        'email[2]': 'asdasd@.com',
        'firstName[1]': 'asdasd',
        'firstName[2]': 'asdasd',
        'lastName[1]': 'asdasd',
        'lastName[2]': 'asd',
        test: 'test',
      }),
    ).toMatchSnapshot();
  });

  it('should combine array object correctly', () => {
    expect(
      combineFieldValues({
        'name[0].firstName': 'testFirst',
        'name[0].lastName': 'testLast',
        'test[1].what': 'testLast',
        'test[1].task': 'testLast',
        'test[2].what': 'testLast',
        'test[2].what[1].test': 'testLast',
      }),
    ).toMatchSnapshot();
  });

  it('should combine object correctly', () => {
    expect(
      combineFieldValues({
        'name.firstName': 'testFirst',
        'name.lastName': 'testLast',
        'name.lastName.bill.luo': 'testLast',
      }),
    ).toMatchSnapshot();
  });

  it('should return default name value', () => {
    expect(
      combineFieldValues({
        'name': 'testFirst',
      }),
    ).toMatchSnapshot();
  });
});
