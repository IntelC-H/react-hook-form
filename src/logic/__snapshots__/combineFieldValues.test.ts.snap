// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combineFieldValues should combine all the array fields 1`] = `
Object {
  "email": Array [
    "asdasd@dsad.com",
    "asdasd@.com",
  ],
  "firstName": Array [
    "asdasd",
    "asdasd",
  ],
  "lastName": Array [
    "asdasd",
    "asd",
  ],
  "test": "test",
}
`;

exports[`combineFieldValues should combine array object correctly 1`] = `
Object {
  "name": Array [
    Object {
      "firstName": "testFirst",
      "lastName": "testLast",
    },
  ],
  "test": Array [
    Object {
      "task": "testLast",
      "what": "testLast",
    },
    Object {
      "what": Array [
        undefined,
        Object {
          "test": "testLast",
        },
      ],
    },
  ],
}
`;

exports[`combineFieldValues should combine object correctly 1`] = `
Object {
  "name": Object {
    "firstName": "testFirst",
    "lastName": Object {
      "bill": Object {
        "luo": "testLast",
      },
    },
  },
}
`;
